From dc53d7490e5d86eaceefa63111e6c5f99d81b91e Mon Sep 17 00:00:00 2001
From: Jouni Malinen <quic_jouni@quicinc.com>
Date: Wed, 3 Jul 2024 15:13:29 +0530
Subject: [PATCH] hostapd: Allow RSNE Override element to override RSNE
contents during parsing

Parse RSNE override element over RSNE IE contents if both IEs exists.

Patch-dependency: None

Signed-off-by: Jouni Malinen <quic_jouni@quicinc.com>
Signed-off-by: Arunpandi Kannan <quic_arunpand@quicinc.com>
---
 src/common/wpa_common.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/common/wpa_common.c b/src/common/wpa_common.c
index c809bce..12e3710 100644
--- a/src/common/wpa_common.c
+++ b/src/common/wpa_common.c
@@ -1804,6 +1804,14 @@ int wpa_parse_wpa_ie_rsn(const u8 *rsn_ie, size_t rsn_ie_len,
 		data->has_group = 1;
 		data->key_mgmt = WPA_KEY_MGMT_OSEN;
 		data->proto = WPA_PROTO_OSEN;
+	} else if (rsn_ie_len >= 2 + 4 + 2 && rsn_ie[1] >= 4 + 2 &&
+		   rsn_ie[1] == rsn_ie_len - 2 &&
+		   (WPA_GET_BE32(&rsn_ie[2]) == RSNE_OVERRIDE_IE_VENDOR_TYPE ||
+		    WPA_GET_BE32(&rsn_ie[2]) ==
+		    RSNE_OVERRIDE_2_IE_VENDOR_TYPE) &&
+		   WPA_GET_LE16(&rsn_ie[2 + 4]) == RSN_VERSION) {
+		pos = rsn_ie + 2 + 4 + 2;
+		left = rsn_ie_len - 2 - 4 - 2;
 	} else {
 		const struct rsn_ie_hdr *hdr;
 
-- 
2.34.1

